pandas/core/indexes/multi.py

```diff
@@ -3235,9 +3235,13 @@ def union(self, other, sort=None):
 
         # TODO: Index.union returns other when `len(self)` is 0.
 
-        uniq_tuples = lib.fast_unique_multiple(
-            [self._values, other._ndarray_values], sort=sort
-        )
+        if not is_object_dtype(other.dtype):
+            raise NotImplementedError(
+                "Can only union MultiIndex with MultiIndex or Index of tuples, "
+                "try mi.to_flat_index().union(other) instead."
+            )
+
+        uniq_tuples = lib.fast_unique_multiple([self._values, other._values], sort=sort)
```

Comment:
By not passing an ndarray (`_ndarray_values`), I think `fast_unique_multiple` might get a lot slower?

Reply:
we only get here with object dtype, so they are equivalent

Reply:
Ah, yes, I had interpreted the `if not is_object_dtype(..)` the other way around ..